pkg/service/model_builder.go

```diff
@@ -80,17 +92,67 @@ type defaultModelBuilder struct {
 	enableBackendSG          bool
 	disableRestrictedSGRules bool
 
+	stackGlobalCache map[core.StackID]core.Stack
+
 	clusterName         string
 	vpcID               string
 	defaultTags         map[string]string
 	externalManagedTags sets.String
 	defaultSSLPolicy    string
 	defaultTargetType   elbv2model.TargetType
 	enableIPTargetType  bool
+
+	client client.Client
+
+	initialized bool
+	lock        sync.Mutex
 }
 
 func (b *defaultModelBuilder) Build(ctx context.Context, service *corev1.Service) (core.Stack, *elbv2model.LoadBalancer, bool, error) {
-	stack := core.NewDefaultStack(core.StackID(k8s.NamespacedName(service)))
+	// Initialize the global cache if not initialized
+	if !b.initialized {
```

Comment:
Since we are worried about this being called from multiple goroutines (based on the fact that you have a `lock`), then I would think that the initialization would also need to be protected by the lock. Otherwise, there may be multiple attempts at initialization.

Reply:
see https://github.com/acorn-io/aws-load-balancer-controller/pull/1#discussion_r1379837062

Now the whole stack deploy/build will be locked to make sure stack is thread safe.